fix for page not rendering on /find-locations/facility/vha_438GE

accbjt opened the pull request:
## Description
Facility Page does not render

Repro:
1/ Go to VA.gov -> Find a Location
2/ Type in "Wagner, SD", select "VA Health", click Search
3/ In the results list click "Wagner, SD VA Clinic"
--> The page renders with header and footer but no body content.
--> https://www.va.gov/find-locations/facility/vha_438GE

According to @ayaleloehr the API is returning complete, valid results. So this appears to be a bug in the tool itself.

## Testing done
Manual

## Screenshots
<img width="339" alt="screen shot 2019-02-14 at 4 55 17 am" src="https://user-images.githubusercontent.com/5377648/52788921-4a461a00-3017-11e9-9b0e-cd68ce0b4591.png">

## Acceptance criteria
- [ ] Page should render

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

accbjt committed:
fix for page not rendering on /find-locations/facility/vha_438GE

accbjt commented:
The problem was with the data coming back. Here is the area in data that is causing an error. If you take a look at **friday**, it is not correctly formatted. I added some code to deal with this but we need to fix it in the backend so that it doesn't give us the wrong format.

```
hours: {
   monday: '800AM-430PM',
   tuesday: '800AM-430PM',
   wednesday: '800AM-1200PM',
   thursday: '800AM-430PM',
   friday: '00AM-30PM',
   saturday: '-',
   sunday: '-',
},
```

This is what it looks like with my fix. Take a look at **friday**. Not sure if this is actually what we want if the time has an error.

![Screen Shot 2019-02-14 at 4.55.17 AM.png](https://images.zenhubusercontent.com/5ac3be554b5806bc2bcd7ea1/13cbe922-5e85-44ad-8dc8-ff53d821d998)

accbjt committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into bug/16722

accbjt committed:
add raven error to track any instance of malformed data

accbjt committed:
removed code not needed

accbjt commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -63,12 +65,22 @@ export default class LocationHours extends Component {
       } else if (hours[k]) {
-        const regex = /(\d+\w\w)-(\d+\w\w)/;
+        const regex = /(([1-8][0-9]{2}|9[0-8][0-9]|99[0-9]|1[01][0-9]{2}|12[0-4][0-9]|125[0-9])+\w\w)-(([1-8][0-9]{2}|9[0-8][0-9]|99[0-9]|1[01][0-9]{2}|12[0-4][0-9]|125[0-9])+\w\w)/;
         const found = hours[k].match(regex);
```
Added this to check the correct format for hours being sent from the API.

accbjt commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -63,12 +65,22 @@ export default class LocationHours extends Component {
-            found[2],
+            found[3],
           )}`;
+        } else {
```
add this so we can track when this happens

accbjt committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into bug/16722

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -63,12 +65,22 @@ export default class LocationHours extends Component {
         mappedHours[k] = 'Closed';
       } else if (hours[k]) {
-        const regex = /(\d+\w\w)-(\d+\w\w)/;
+        const regex = /(([1-8][0-9]{2}|9[0-8][0-9]|99[0-9]|1[01][0-9]{2}|12[0-4][0-9]|125[0-9])+\w\w)-(([1-8][0-9]{2}|9[0-8][0-9]|99[0-9]|1[01][0-9]{2}|12[0-4][0-9]|125[0-9])+\w\w)/;
```
This is pretty rough to understand. Also, the parsing in this component is just pretty brittle in general.

I think we could get to a simpler, more reliable place by using `moment`. You could split the date string at `-` and then parse each piece using `moment(timeString, "hmma")`. You can then run `isValid()` on the result to see if it parsed correctly. And instead of using the `colonizeTime` function, you can format it like `momentDate.format("h:mma")`.

accbjt replied:
@jbalboni Yeah I didn't really like how this was written. I will do some refactoring so it's better to read and more reliable.

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -63,12 +65,22 @@ export default class LocationHours extends Component {
           )}`;
+        } else {
+          Raven.captureException('API data malformed', {
+            data: this.props.location,
```
I think you need to wrap `data` and `errorData` in an `extra` object for Sentry to see it properly. https://docs.sentry.io/clients/javascript/usage/#raven-js-additional-context

accbjt replied:
Thanks

accbjt committed:
add test for location hours

accbjt committed:
add test and refactored LocationHours component

accbjt committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into bug/16722

accbjt commented:
I refactored the code. Let me know what you think. I added some tests to make sure my refactor doesn't effect anything. We are now using moment to check if time data is correctly formatted. For some reason I could not get it to give me back `AM and PM` in capitals. Here is what it looks like now on the page, which I think looks okay.

<img width="324" alt="screen shot 2019-02-20 at 4 55 44 pm" src="https://user-images.githubusercontent.com/5377648/53137214-b86a6f80-3536-11e9-8b10-5f53292e36ac.png">

accbjt committed:
fix name of method in test for LocationHours

accbjt committed:
change data for mock api to valid data

accbjt commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
+    const hours = hour.split('-');
+    const isValid = hours.every(time => moment(time, 'hmmA').isValid());
+
+    if (!isValid) {
```
@jbalboni The extra property in  Raven.captureMessage object was an example I found on how to add extra data to the error log. I'm not totally sure if it works.  Here is where I got it from. Is there a way to check if this got recorded in Sentry?

https://github.com/getsentry/sentry-javascript/issues/167

accbjt commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
+    const isValid = hours.every(time => moment(time, 'hmmA').isValid());
+
+    if (!isValid) {
+      Raven.captureMessage('API data malformed', {
```
@jbalboni The extra property in  Raven.captureMessage object was an example I found on how to add extra data to the error log. I'm not totally sure if it works.  Here is where I got it from. Is there a way to check if this got recorded in Sentry?

https://github.com/getsentry/sentry-javascript/issues/167

accbjt commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
+
+    if (!isValid) {
+      Raven.captureMessage('API data malformed', {
+        extra: {
```
@jbalboni The extra property in  Raven.captureMessage object was an example I found on how to add extra data to the error log. I'm not totally sure if it works.  Here is where I got it from. Is there a way to check if this got recorded in Sentry?

https://github.com/getsentry/sentry-javascript/issues/167

jbalboni replied:
There's not a way locally, but we can test it in staging, if we can find bad data.

Getting things correct in Sentry is kinda trial and error before it goes live; if have to wait to check this in prod, that's ok.

jbalboni commented:
@accbjt I believe that's actually the desired formatting of a.m./p.m. per the content folks, so you're good there.

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
-    return `${found[1]}:${found[2]}${found[3]}`;
+  convertHour(hour, day) {
+    const hours = hour.split('-');
+    const isValid = hours.every(time => moment(time, 'hmmA').isValid());
```
You could get rid of the extra parsing iteration (here and the bottom of the function) by doing `const hours = hour.split('-').map(time => moment(time, 'hmmA'));` and using the array of moment objects on this line and in the later return statement.

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
+    }, {});
+  }
+
+  isValidObject(location) {
```
This is a sort of generic function name, too; what about `isValidData` or `isLocationDataValid`?

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
-  colonizeTime(time) {
-    const found = time.match(/(\d?\d)(\d\d)(\w\w)/);
-    return `${found[1]}:${found[2]}${found[3]}`;
+  convertHour(hour, day) {
```
This function name is pretty generic, what about `formatTimeRange`?

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
-          </div>
-        </div>
-      );
+  convertHours(hours) {
```
Same as above, I think this function name is pretty broad. `formatLocationHours`, maybe?

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
+        extra: {
+          data: this.props.location,
+          errorData: {
+            hours: { [day]: hour },
```
You could probably get away with just including the location data and making the error message say that it's related to hours. Anyone looking at the sentry error should be able to figure out the error just by looking at the data or by putting it in the mock data. This would let you remove the `day` parameter in this function.

jbalboni commented on src/applications/facility-locator/components/LocationHours.jsx:
```diff
@@ -1,27 +1,61 @@
+        return { ...accum };
+      }
+
+      return { ...accum, [key]: this.convertHour(hours[key]) };
```
Missing the second argument here, but see above comment about maybe removing that.

jbalboni reviewed (commented):
This is vastly better than what was there before. I added some minor comments around readability.

accbjt committed:
update according to revision request

accbjt committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into bug/16722

accbjt commented:
@jbalboni This should be ready to go.
